### ISSUE TITLE
cell editing quirk in TableBrowser (doubleclick vs two clicks vs F2)

### ISSUE THREAD

**Issue by hertzhaft** [sqlitebrowser/sqlitebrowser#2293]
Details for the issue
--------------------

I want to edit a table cell. I click on it: The cell gets *selected*. I wait a little and click a second time: The cell becomes editable (has a thicker border and a cursor) - but only after another short delay.

versus:

I double-click on the cell: The cell stays *selected* and does not become editable.

versus:

I press `[F2]` when the cell is *selected*, and it becomes immediately editable (without delay).

#### What did you expect to see?

The cell to become immediately editable in all three cases.

#### What did you see instead?

Two things that will confuse/annoy users (in my opinion): 
- the pause that is needed between the two clicks (double-clicking having no effect)
- the delay after the second click before the cell becomes editable

#### What operating system are you using?

- [X] Linux: Arch Linux 5.7.4-arch1-1 

#### What is your DB4S version?

- [X] Other:  3.12.99 (Jun 13 2020)

#### Did you also

- [ ] Try out the latest nightly version: https://github.com/sqlitebrowser/sqlitebrowser#nightly-builds
- [X] Search for an existing similar issue: https://github.com/sqlitebrowser/sqlitebrowser/issues?utf8=%E2%9C%93&q=is%3Aissue%20

**Comment by justinclift** [comment]
In theory, when you double click on the cell that should open it in the Edit Database Cell pane.

Personally, I find that can get in the way more than it helps sometimes though (eg when tabbing through fields).

**Comment by chrisjlocke** [comment]
Is this with the 'Edit Database Cell' dock window open?  Usually, double-clicking a cell shifts focus to this dock window with the contents of the cell in the cell editor.

**Comment by mgrojo** [comment]
Yes, double-clicking the cell will change focus to the Edit Dabase Cell pane. I will open the window if it is hidden, so when you say that double-clicking have no effect, I suppose you mean if your aim is to edit the cell inline. Otherwise, you're seeing a bug, which I cannot reproduce.

I don't think we should change the current behaviour, because a lot of users, including me, is used to that. All programs need some discovery or directed training to be used by new users.

**Comment by hertzhaft** [comment]
I have to correct myself: double-clicking in fact shifts focus to the Edit Database Cell pane. I just wasn't aware of it: There is no blinking cursor (which I would expect), just a slight change of the background color (which I overlooked - I didn't recognize that as an indication that the pane becomes editable).

Now I see that for JSON mode the cursor *does* appear in the Edit Database Cell pane. For text mode it doesn't. Hadn't noticed that before.

Of course you can get used to the way it works now, but I'd still say it's somewhat inconsistent :wink: ...

Any opinions on the short delay when clicking twice (with a pause)?

**Comment by chrisjlocke** [comment]
> Any opinions on the short delay when clicking twice (with a pause)?

Is this the half a second delay after releasing the mouse button, and *then* it goes into edit mode?  I'm guessing (although I have no way of checking) that this is entirely down to the control itself, and is not a variable that we can manage (ie, reduce the delay).  

I'll add a Wiki article on the click behaviour so at least its documented.  

With regard to the 'double-clicking in text mode doesn't show a cursor', it does, but as its right at the start and the text is highlighted, it's hard to see (this is on Windows OS though, so not 10000000000000% sure its the same on Linux).

**Comment by mgrojo** [comment]
> > Any opinions on the short delay when clicking twice (with a pause)?
> 
> Is this the half a second delay after releasing the mouse button, and _then_ it goes into edit mode? I'm guessing (although I have no way of checking) that this is entirely down to the control itself, and is not a variable that we can manage (ie, reduce the delay).

Yes, this is provided by QTableView and as far as I know, the delay would be controlled by the platform maximum delay configured for double-click: when it is longer than that delay, it is a single click of the selected cell. 

> 
> I'll add a Wiki article on the click behaviour so at least its documented.

Awesome.

> With regard to the 'double-clicking in text mode doesn't show a cursor', it does, but as its right at the start and the text is highlighted, it's hard to see (this is on Windows OS though, so not 10000000000000% sure its the same on Linux).

Maybe the cursor should be at the end of the text like in version 3.11 and the RTL Text mode? In general, I see that modern desktop styles does not provide good feedback for the focused widget.

**Comment by chrisjlocke** [comment]
> and as far as I know, the delay would be controlled by the platform maximum delay configured for double-click

Just to confirm, this isn't a delay to determine a double click but is the delay after a single click.  I guess you mean it has to wait for this delay in case you click again during this time to mean a double-click.  Aah, makes sense, but in practice, that's annoying! 😉   
Could be useful to have an option for 'aah, you mean you want to inline-edit instead' as that's a semi-expected behaviour.  A crude hack for this could be simulating an F2 keystroke on a double-click, but there might be a programmatically better way....

**Comment by mgrojo** [comment]
> > and as far as I know, the delay would be controlled by the platform maximum delay configured for double-click
> 
> Just to confirm, this isn't a delay to determine a double click but is the delay after a single click. I guess you mean it has to wait for this delay in case you click again during this time to mean a double-click. Aah, makes sense, but in practice, that's annoying! wink
> Could be useful to have an option for 'aah, you mean you want to inline-edit instead' as that's a semi-expected behaviour. A crude hack for this could be simulating an F2 keystroke on a double-click, but there might be a programmatically better way....

I'm not sure I'm following you.

The point is if we have a double click or a succession of two single clicks. If the time between the two clicks is under the OS maximum time for a double click, a double click is signalled, and we open and focus the "Edit Database Cell" dock window. If the time is greater than the double-click maximum time, two single clicks are signalled, the first one selects the cell and the second one activates the edition. No matter how long separated are the clicks, it's binary: either it's inside the time, and a double-click is signalled, or it's greater than that time and two single clicks are signalled.

This is fairly standard. You get the same behaviour in file icons in Windows: one double-click opens the file; two single clicks edits the filename. In my opinion, the only thing that we could do is adding both actions to the contextual menu, which is what you also get in Windows. The drawback: more entries in the menu for something that most people will do using the shortcuts, if they know it, so maybe it's better to just train them. But I don't strongly oppose the menu entries.

**Comment by chrisjlocke** [comment]
> I'm not sure I'm following you.

Apologies - I waffle too much!

> This is fairly standard. You get the same behaviour in file icons in Windows

Exactly this.  I'm sure if you timed it, right-clicking and selecting 'Edit' would be the same as single-clicking, waiting, and single-clicking again... it just seems a year waiting rather than doing something, so you are correct.

**Comment by hertzhaft** [comment]
I see the rationale now. I guess the thing that confused me was that I wasn't aware that the focus had shifted when I double-clicked the cell. So setting the cursor to the end of the Edit Database Cell pane  (as it was in 3.11.2, I notice it just now) would perhaps be of help, and perhaps dimming the selection background of the table cell would also make it more clear to the user that the focus has shifted. A context menu entry to me is not needed, in my opinion. Thanks for the clarification!
